drivers: video: ov7725: add sensor driver for ov7725

lgl88911 opened the pull request:
This patch adds the driver for Omnivision OV7725
Color CMOS VGA Sensor.

The driver currently provides a simple capture
function, the output format only provides
RGB565,640x480.

Signed-off-by: Frank Li <lgl88911@163.com>

loicpoulain commented:
It looks interesting, but could you please split it in several commits, at least one for the dts changes, one for the new ov driver and one for the sample.

lgl88911 commented:
> @lgl88911, It looks interesting, but could you please split it in several commits, at least one for the dts changes, one for the new ov driver and one for the sample.
Thanks for review, the PR will be close and I divided this commit into three: 
1. OV7725 driver https://github.com/zephyrproject-rtos/zephyr/pull/30787
In order to avoid CI fail, After https://github.com/zephyrproject-rtos/zephyr/pull/30787 be merged, the following two branches will make a pull request.
2. mm_swiftio support OV7725
3. sample

loicpoulain commented:
> > @lgl88911, It looks interesting, but could you please split it in several commits, at least one for the dts changes, one for the new ov driver and one for the sample.
> > Thanks for review, the PR will be close and I divided this commit into three:
> 
>     1. OV7725 driver #30787
>        In order to avoid CI fail, After #30787 be merged, the following two branches will make a pull request.
> 
>     2. mm_swiftio support OV7725, https://github.com/lgl88911/zephyr/tree/mm_swiftio_ov7725
> 
>     3. sample，https://github.com/lgl88911/zephyr/tree/add_ov7725_sample

No, keep one branch and one PR/MR, don't need to close this one, I just wanted you split your changes in multiple git commits not in multiple branch.

lgl88911 committed:
drivers: video: ov7725: add sensor driver for ov7725

This patch adds the driver for Omnivision OV7725
Color CMOS VGA Sensor.

The driver currently provides a simple capture
function, the output format only provides
RGB565,640x480.

Signed-off-by: Frank Li <lgl88911@163.com>

lgl88911 commented:
@loicpoulain Thanks for your explain, I split it to three commit, Please help review again

loicpoulain commented on dts/bindings/vendor-prefixes.txt:
```diff
@@ -306,6 +306,7 @@ okaya	Okaya Electric America, Inc.
 oki	Oki Electric Industry Co., Ltd.
 olimex	OLIMEX Ltd.
 olpc	One Laptop Per Child
+omnivision OmniVision Technologies Co., Ltd.
```
Under Linux OmniVision Technologies vendor prefix is 'ovti' , I would prefer that one for consistency.

lgl88911 replied:
Very reasonable. Fixed

loicpoulain commented on dts/bindings/video/omnivision,ov7725.yaml:
```diff
@@ -0,0 +1,16 @@
+description: OV7725 CMOS video sensor
+
+compatible: "omnivision,ov7725"
+
```
ovti,ov7725

lgl88911 replied:
Very reasonable , fFxed

loicpoulain commented on drivers/video/ov7725.c:
```diff
@@ -0,0 +1,651 @@
+	return ov7725_init(dev);
+}
+
+DEVICE_AND_API_INIT(ov7725, DT_INST_LABEL(0), &ov7725_init_0,
```
That needs to be converted to the new DT device macro: DEVICE_AND_API_INIT -> DEVICE_DT_INST_DEFINE
cf https://github.com/zephyrproject-rtos/zephyr/pull/30779/commits/fc110b67ea66c77c43565a014583ab00012ff409

lgl88911 replied:
Change is so fast @@. converted done.

loicpoulain reviewed (changes requested):
Overall, it looks good, just few comments to address.

loicpoulain commented on dts/bindings/video/ovti,ov7725.yaml:
```diff
@@ -0,0 +1,16 @@
+properties:
+    reset-gpios:
+      type: phandle-array
+      required: false
```
Is it really optional? the driver seems to unconditionnaly request it.

lgl88911 replied:
Yes, If reset is connected to the reset circuit of the hardware, there is no need to use GPIO to control it.
If the gpio name cannot be found in the driver, gpio reset will not be performed
```
char *gpio_name = DT_INST_GPIO_LABEL(0, reset_gpios);

if (gpio_name) {
		drv_data->reset_gpio = device_get_binding(gpio_name);
		.....
}

...

if (drv_data->reset_gpio) {
		ret = gpio_pin_configure(drv_data->reset_gpio,
					 drv_data->reset_pin,
					 GPIO_OUTPUT_ACTIVE |
					 drv_data->reset_flags);
		if (ret) {
			return ret;
		}

		gpio_pin_set(drv_data->reset_gpio,
			     drv_data->reset_pin,
			     0);
		k_sleep(K_MSEC(1));
		gpio_pin_set(drv_data->reset_gpio,
			     drv_data->reset_pin,
			     1);
		k_sleep(K_MSEC(1));
	}
```

MaureenHelm commented on drivers/video/ov7725.c:
```diff
@@ -0,0 +1,651 @@
+#define OV7725_SIGN       0xABU
+#define OV7725_DSPAUTO    0xACU
+
+#define OV7725_COM10_VSYNC_NEG_MASK    (1U << 1U)
+#define OV7725_COM10_HREF_REVERSE_MASK (1U << 3U)
+#define OV7725_COM10_PCLK_REVERSE_MASK (1U << 4U)
+#define OV7725_COM10_PCLK_OUT_MASK     (1U << 5U)
+#define OV7725_COM10_DATA_NEG_MASK     (1U << 7U)
```
Recommend using `BIT()` from `include/sys/util.h`

lgl88911 replied:
Thanks for review, Fixed

MaureenHelm commented on drivers/video/ov7725.c:
```diff
@@ -0,0 +1,651 @@
+	{ .pixel_format = VIDEO_PIX_FMT_RGB565, .com7 = (1 << 2) | (2) }
+};
+
+static struct ov7725_reg ov7725_init_reg_tb[] = {
```
`const`?

lgl88911 replied:
Yes, lost const, Fixed

MaureenHelm commented on drivers/video/ov7725.c:
```diff
@@ -0,0 +1,651 @@
+	msgs[1].len = 1;
+	msgs[1].flags = I2C_MSG_WRITE | I2C_MSG_STOP;
+
+	return i2c_transfer(drv_data->i2c, msgs, 2, drv_data->i2c_addr);
```
Is there a reason you can't use `i2c_reg_write_byte()` instead?

lgl88911 replied:
Although i2c_reg_write_byte can achieve the same purpose, I hope it is the similar as ov7725_read_reg.

MaureenHelm commented on drivers/video/ov7725.c:
```diff
@@ -0,0 +1,651 @@
+	msgs[1].len = 1;
+	msgs[1].flags = I2C_MSG_READ | I2C_MSG_STOP | I2C_MSG_RESTART;
+
+	return i2c_transfer(drv_data->i2c, msgs, 2, drv_data->i2c_addr);
```
Replace with `i2c_reg_read_byte()`

lgl88911 replied:
OV7725 uses SCCB to communicate. SCCB is similar to I2C, but not exactly the same. When SCCB reads the register, an I2C stop bit is required after the register address of OV7725 is written, and then the read operation can be performed. i2c_reg_read_byte call i2c_write_read, Its implementation lacks a stop bit
```
{
	struct i2c_msg msg[2];

	msg[0].buf = (uint8_t *)write_buf;
	msg[0].len = num_write;
	msg[0].flags = I2C_MSG_WRITE;

	msg[1].buf = (uint8_t *)read_buf;
	msg[1].len = num_read;
	msg[1].flags = I2C_MSG_RESTART | I2C_MSG_READ | I2C_MSG_STOP;

	return i2c_transfer(dev, msg, 2, addr);
}
```

loicpoulain replied:
Would be nice to add the SCCB explanation as an inline comment.

lgl88911 replied:
Thanks for suggest, Added it.

MaureenHelm commented on drivers/video/ov7725.c:
```diff
@@ -0,0 +1,651 @@
+	int ret;
+	uint8_t set_value;
+
+	ret = ov7725_read_reg(dev, reg_addr, &set_value);
+
+	if (ret == 0) {
+		set_value = (set_value & (~clear_mask)) |
+				(set_value & clear_mask);
+		ret = ov7725_write_reg(dev, reg_addr, set_value);
+	}
```
Replace with `i2c_reg_update_byte()`

lgl88911 replied:
The reason is similar to ov7725_read_reg, i2c_reg_update_byte will fail when reading SCCB.

MaureenHelm commented on boards/arm/mm_swiftio/pinmux.c:
```diff
@@ -20,30 +20,30 @@ static gpio_pin_config_t enet_gpio_config = {
 	uint32_t speed, uint32_t strength)
 {
 	uint32_t cmd_data = IOMUXC_SW_PAD_CTL_PAD_SPEED(speed) |
-				IOMUXC_SW_PAD_CTL_PAD_SRE_MASK |
```
Please move the unrelated whitespace changes to a separate commit

lgl88911 replied:
Thanks for review, Fixed

MaureenHelm commented on samples/video/capture/boards/mm_swiftio.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_VIDEO_OV7725=y
+CONFIG_I2C=y
```
Please add this board to `samples/video/capture/sample.yaml` to make sure we build it in CI

lgl88911 replied:
I didn’t find any documentation about sample.yaml. I added one by referring to the existing example. It doesn’t seem to cause CI check fail, but I’m not sure that CI checked mm_swiftio or not. Please help to review whether this modification is correct. Thanks so much!

MaureenHelm reviewed (changes requested):
Thanks @lgl88911 !

lgl88911 committed:
boards: mm_swiftio: modify to follow code style

Move the unrelated whitespace and
Modify line break alignment.

Signed-off-by: Frank Li <lgl88911@163.com>

lgl88911 committed:
boards: mm_swiftio: add camera support

Add camera support, use OV7725 CMOS sensor.

Signed-off-by: Frank Li <lgl88911@163.com>

lgl88911 committed:
samples: video: Add OV7725 video capture sample

Add the ov7725 senser example,
which can run normally on mm_swiftio.

Signed-off-by: Frank Li <lgl88911@163.com>

loicpoulain commented on samples/video/capture/sample.yaml:
```diff
@@ -6,3 +6,8 @@ tests:
     tags: video
     platform_allow: mimxrt1064_evk
```
That is a list of allowed boards so you can directly append mm_swiftio here.

lgl88911 replied:
Thanks so much, fixed

lgl88911 commented:
@MaureenHelm  Can you help review again,Thanks
